File: src/domain/usecases/support/SupportTicketUseCases.ts
import type { SupportTicket } from '@/domain/entities/SupportTicket'
import type { User } from '@/domain/entities/User'
import { assertUserCanManageUsers } from '@/domain/entities/User'
import type { SupportTicketRepository } from '@/domain/repositories/SupportTicketRepository'
import {
  UnauthorizedError,
  ValidationError,
} from '@/domain/errors/DomainError'

export class ListSupportTicketsUseCase {
  private readonly repository: SupportTicketRepository

  constructor(repository: SupportTicketRepository) {
    this.repository = repository
  }

  async execute(actor: User): Promise<SupportTicket[]> {
    if (!assertUserCanManageUsers(actor)) {
      throw new UnauthorizedError('Solo el administrador puede ver el soporte')
    }
    return this.repository.listAll()
  }
}

export class ResolveSupportTicketUseCase {
  private readonly repository: SupportTicketRepository

  constructor(repository: SupportTicketRepository) {
    this.repository = repository
  }

  async execute(
    actor: User,
    input: { ticketId: string; response: string },
  ): Promise<SupportTicket> {
    if (!assertUserCanManageUsers(actor)) {
      throw new UnauthorizedError('Solo el administrador puede responder')
    }
    const ticketId = input.ticketId.trim()
    if (!ticketId) throw new ValidationError('Aviso inválido')
    const response = input.response.trim()
    if (response.length > 1000) {
      throw new ValidationError('La respuesta es demasiado larga')
    }
    return this.repository.resolve({
      ticketId,
      response,
      resolvedById: actor.id,
      resolvedByName: actor.displayName,
    })
  }
}

File: consorcio/lib/domain/usecases/support_ticket_use_cases.dart
import '../entities/app_user.dart';
import '../entities/support_ticket.dart';
import '../errors/domain_exception.dart';
import '../repositories/support_ticket_repository.dart';

const _kinds = {'SUGERENCIA', 'PROBLEMA'};

class CreateSupportTicketUseCase {
  CreateSupportTicketUseCase(this._repository);

  final SupportTicketRepository _repository;

  Future<SupportTicket> execute(
    AppUser actor, {
    required String kind,
    required String message,
  }) async {
    actor.assertCanOperateApp();
    final nextKind = kind.trim().toUpperCase();
    if (!_kinds.contains(nextKind)) {
      throw DomainException('Elige si es sugerencia o problema');
    }
    final text = message.trim();
    if (text.length < 8) {
      throw DomainException('Describe el aviso con al menos 8 caracteres');
    }
    if (text.length > 1000) {
      throw DomainException('El aviso es demasiado largo');
    }
    return _repository.create(
      kind: nextKind,
      message: text,
      createdById: actor.id,
      createdByName: actor.displayName,
    );
  }
}

class ListSupportTicketsUseCase {
  ListSupportTicketsUseCase(this._repository);

  final SupportTicketRepository _repository;

  Future<List<SupportTicket>> execute(AppUser actor) async {
    actor.assertCanOperateApp();
    if (actor.isMobileAdmin) {
      return _repository.listAll();
    }
    return _repository.listMine(actor.id);
  }
}

class ResolveSupportTicketUseCase {
  ResolveSupportTicketUseCase(this._repository);

  final SupportTicketRepository _repository;

  Future<SupportTicket> execute(
    AppUser actor, {
    required String ticketId,
    required String response,
  }) async {
    actor.assertCanOperateApp();
    if (!actor.isMobileAdmin) {
      throw DomainException('Solo el administrador puede responder el aviso');
    }
    final text = response.trim();
    if (text.length > 1000) {
      throw DomainException('La respuesta es demasiado larga');
    }
    return _repository.resolve(
      ticketId: ticketId.trim(),
      response: text,
      resolvedById: actor.id,
      resolvedByName: actor.displayName,
    );
  }
}
